Mathlib/Algebra/Group/Irreducible.lean

```diff
@@ -0,0 +1,76 @@
+/-
+Copyright (c) 2018 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl, Jens Wagemaker, Yaël Dillies
+-/
+import Mathlib.Algebra.Divisibility.Basic
+import Mathlib.Algebra.Group.Units.Basic
+
+/-!
+# Irreducible elements in a monoid
+
+This file defines irreducible elements of a monoid, as non-units that can't be written as the sum of
+two non-units. This generalises irreducible elements of a ring.
+
+In decomposition monoids (e.g., `ℕ`, `ℤ`), this predicate is equivalent to `Prime`
+(see `irreducible_iff_prime`), however this is not true in general.
+-/
+
+assert_not_exists MonoidWithZero OrderedCommMonoid Multiset
+
+variable {M : Type*}
+
+/-- `Irreducible p` states that `p` is non-unit and only factors into units.
+
+We explicitly avoid stating that `p` is non-zero, this would require a semiring. Assuming only a
+monoid allows us to reuse irreducible for associated elements. -/
```

Comment:
I don't think this paragraph really works in the additive case. Otherwise seems fine